Usecase test 1: demonstrating that genes given in an article's abstract are found to be amoung the highest significance genes linked to that article in the KG. 

1. Retrieving all named genes from the 10 articles first retrieved from Pubmed (from the stored metadata file)

In [9]:
#import scispacy
import spacy
import csv
import re
spacy.load("en_core_web_sm")
#from spacy.lang.en import English
#retrieve data from the metadata file
def get_csv_column(file_path, column_name):
    data = []
    with open(file_path, 'r') as csvfile:
        csvreader = csv.DictReader(csvfile)
        for row in csvreader:
            data.append(row[column_name])
    return data


def process_csv(file_path, search_terms, search_column, return_column):
    nlp = spacy.load("en_core_web_sm")
    results = []
    search_terms_lower = [term.lower() for term in search_terms]
    with open(file_path, 'r', encoding='utf-8') as csvfile:
        reader = csv.DictReader(csvfile)
        for row in reader:
            search_text = row[search_column]
            return_value = row[return_column]
            doc = nlp(search_text)
            matched_terms = set(token.text for token in doc 
                                if token.text.lower() in search_terms_lower 
                                and not token.is_punct
                                and token.pos_ != "VERB")
            if matched_terms:
                results.append({
                    'matched_terms': list(matched_terms),
                    'return_value': return_value
                })
    return results


abstract_file = 'data/asd_article_metadata.csv'
gene_file = 'gene_list.csv'     #list of gene names taken from Ensembl 
column_name = 'Gene name'
gene_list = get_csv_column(gene_file, column_name)
search_column = 'abstract'      #will search through the columns of abstracts
return_column = 'pmid'           #will return the associated article pmid

results = process_csv(abstract_file, gene_list, search_column, return_column)

for result in results:
    print(f"Article PMID: {result['return_value']}")
    print(f"Matched gene terms: {result['matched_terms']}")
    
    print()

Article PMID: 36213201
Matched gene terms: ['set']

Article PMID: 36688057
Matched gene terms: ['CD14']

Article PMID: 37381037
Matched gene terms: ['ATRX', 'set', 'TEs', 'impact']

Article PMID: 38781372
Matched gene terms: ['GRN']



37381037 - contains ATRX


2. Querying the KG to retrieve the most significant genes within the datasets for each article: 

In [3]:
import rdflib
#from rdflib.plugins.sparql import prepareQuery
import requests

filename = "test_graph1.ttl"
g = rdflib.Graph()
g.parse(filename)
#print(len(g))
test_query = """
    PREFIX EDAM: <http://edamontology.org/>
    PREFIX RDF: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX DCT: <http://purl.org/dc/terms/>
    PREFIX BIOLINK: <https://w3id.org/biolink/vocab/>
    
    SELECT DISTINCT ?pmid ?gene ?p_output
    WHERE {
        ?pmid (<>|!<>)* ?gene .
        ?pmid (<>|!<>)* ?p_output .
        ?pmid RDF:type DCT:identifier .
        ?subject ?pvalue ?p_output .
        FILTER (xsd:float(?p_output) < 0.05)  
    }
    """
results = g.query(test_query)
for row in results:
    print(row)
print(len(results))

#        FILTER(REGEX(?pvalue), "p[-\\s]?value|pvalue") .
#        FILTER (xsd:float(?p_output) < 0.05)  
#   FILTER regex(?pvalue, "pvalue", "i")
#    ORDER BY ASC(?p_output)


(rdflib.term.URIRef('https://www.ncbi.nlm.nih.gov/pmc/articles/PMID30545856'), rdflib.term.URIRef('https://www.ncbi.nlm.nih.gov/pmc/articles/PMID30545856'), rdflib.term.Literal('-3.49'))
(rdflib.term.URIRef('https://www.ncbi.nlm.nih.gov/pmc/articles/PMID30545856'), rdflib.term.URIRef('http://purl.org/dc/terms/identifier'), rdflib.term.Literal('-3.49'))
(rdflib.term.URIRef('https://www.ncbi.nlm.nih.gov/pmc/articles/PMID30545856'), rdflib.term.BNode('nb9dd496b2e164bc0925dd30b465e8fddb1'), rdflib.term.Literal('-3.49'))
(rdflib.term.URIRef('https://www.ncbi.nlm.nih.gov/pmc/articles/PMID30545856'), rdflib.term.URIRef('https://w3id.org/biolink/vocab/Dataset'), rdflib.term.Literal('-3.49'))
(rdflib.term.URIRef('https://www.ncbi.nlm.nih.gov/pmc/articles/PMID30545856'), rdflib.term.BNode('nb9dd496b2e164bc0925dd30b465e8fddb2'), rdflib.term.Literal('-3.49'))
(rdflib.term.URIRef('https://www.ncbi.nlm.nih.gov/pmc/articles/PMID30545856'), rdflib.term.Literal('RP11-184A2.2'), rdflib.term.Literal('-3.

KeyboardInterrupt: 